src/Workspaces/Remote/Core/Services/AssetService.cs

```diff
@@ -86,23 +88,67 @@ public void UnregisterAssetSource(int serviceId)
 
         public async Task<T> GetAssetAsync<T>(Checksum checksum, CancellationToken cancellationToken)
         {
+            T asset;
+            if (TryGetAsset(checksum, out asset))
+            {
+                return asset;
+            }
+
             using (Logger.LogBlock(FunctionId.AssetService_GetAssetAsync, GetChecksumLogInfo, checksum, cancellationToken))
             {
+                // TODO: what happen if service doesn't come back. timeout?
+                var value = await RequestAssetAsync(checksum, cancellationToken).ConfigureAwait(false);
+
+                _assets.TryAdd(checksum, new Entry(value));
+
+                return (T)value;
+            }
+        }
+
+        public async Task SynchronizeSolutionAssetsAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
+        {
+            using (Logger.LogBlock(FunctionId.AssetService_SynchronizeSolutionAssetsAsync, GetChecksumLogInfo, solutionChecksum, cancellationToken))
+            {
+                var collector = new ChecksumSynchronizer(this);
+                await collector.SynchronizeAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private bool Exists(Checksum checksum)
```

Comment:
You haven't documented why that's needed.  I figured it out from looking through all the code.  But it wasn't immediately clear why someone would need to ask for this.  It's also unclear that this ends up changing the caching behavior.